sklearn/metrics/classification.py

```diff
@@ -391,72 +392,264 @@ def jaccard_similarity_score(y_true, y_pred, normalize=True,
+                                                 assume_unique=True)])
+
     if y_type.startswith('multilabel'):
+        if average == 'samples':
+            if normalize is None:
+                normalize = True
+        elif normalize is not None:
+            raise ValueError("'normalize' is only meaningful with "
+                             "`average='samples'`, got `average='%s'`."
+                             % average)
+
+        if not np.all(labels == present_labels):
+            if np.max(labels) > np.max(present_labels):
+                raise ValueError('All labels must be in [0, n, labels). '
+                                 'Got %d > %d' %
+                                 (np.max(labels), np.max(present_labels)))
+            if np.min(labels) < 0:
+                raise ValueError('All labels must be in [0, n, labels). '
+                                 'Got %d < 0' % np.min(labels))
+
+        if n_labels is not None:
+            y_true = y_true[:, labels[:n_labels]]
+            y_pred = y_pred[:, labels[:n_labels]]
+
         with np.errstate(divide='ignore', invalid='ignore'):
-            # oddly, we may get an "invalid" rather than a "divide" error here
-            pred_or_true = count_nonzero(y_true + y_pred, axis=1)
-            pred_and_true = count_nonzero(y_true.multiply(y_pred), axis=1)
+
+            if average == 'samples':
+                sum_axis = 1
+                class_weight = sample_weight
+                weights = None
+            elif average == 'micro':
+                sum_axis = 1
+                class_weight = None
+                weights = sample_weight
+            elif average == 'macro':
+                sum_axis = 0
+                class_weight = None
+                weights = sample_weight
+            elif average == 'weighted':
+                sum_axis = 0
+                if sample_weight is None:
+                    class_weight = y_true.toarray().sum(axis=0)
+                else:
+                    class_weight = (y_true.toarray().T).dot(sample_weight)
+                weights = sample_weight
+                if class_weight.sum() == 0:
+                    return 0
+            else:
+                sum_axis = 0
+                weights = sample_weight
+
+            pred_or_true = count_nonzero(y_true + y_pred, axis=sum_axis,
+                                         sample_weight=weights)
+            pred_and_true = count_nonzero(y_true.multiply(y_pred),
+                                          axis=sum_axis,
+                                          sample_weight=weights)
+            if average == 'micro':
+                pred_or_true = np.array([pred_or_true.sum()])
+                pred_and_true = np.array([pred_and_true.sum()])
+
             score = pred_and_true / pred_or_true
             score[pred_or_true == 0.0] = 1.0
+
+            if average is not None:
+                if normalize == False:
+                    if class_weight is not None:
```

Comment:
lgtm.com complains that class_weight here may not have been initialised. This is only the case because it can't identify that the "else" above corresponds to the opposite of "average is not None" guarding this code. So you don't need to fix it, but one way to fix it would be to set class_weight = None before the `if average ==` cases above.